### DISCUSSION TITLE
"Shared storage migration" message won''t go away, doesn't work

### DISCUSSION THREAD

**Discussion by jfburke3** [osmandapp/OsmAnd#15348] [Q&A]
I have a new Android phone and am trying to get OsmAnd (paid version, nightly build) working on it.  Every time I launch the app, the Shared storage migration screen shows up, and I can't get it to go away and _stay_ away.  I used Samsung's SmartSync app to migrate to the new phone.

If I tap Skip, I can proceed into the app.  My favories appear to have been migrated, but none of my recorded tracks, a/v notes, maps, etc., are there.  It appears that I can use the app mostly normally at this point, download maps, etc.  However, I shouldn't **need** to download the maps again, and still can't find my tracks, notes, etc.  If I go into My Places,  then back out, I get the Shared storage migration message again.  If I back out of **any** screen in OsmAnd, I get the Shared storage migration message again.

If I tap Continue, OsmAnd prompts me to select a folder to use (as the destination for data migration?  This isn't clear on the screen).  If I select an existing one OR create a new one and select Use this folder, it prompts me to allow access to it (EVERY time I do this), then goes back to the Shared storage migration screen with a bubble at the bottom that says, "No OsmAnd data found in the selected folder.  Please  try .."  Please try what?  (If I rotate my phone to landscape, the last part is "Please try again."

There doesn't appear to  be anything on the OsmAnd website under Troubleshooting that directly  addresses the problem I'm having.  While there is some stuff on  https://osmand.net/docs/user/troubleshooting/maps-data/ it doesn't tell me how to fix **this** problem (and, the way it is written, doesn't address what happened to my missing tracks/notes/maps and if following some of the steps there will bring them back).

Before I open an Issue over this, does anyone have any suggestions I can try to get OsmAnd working and my data back?

Samsung Galaxy S22 Ultra, Android 12, OsmAnd build 4.3.0#34599m, released 2022-09-16.

Edited to add:

On both old and new phones, the storage path is set to /storage/2751-0650/Android/data/net.osmand.plus/files

On the old phone, that's set for the "External stoarage 2" selection.  On the new phone, it's the "Manually specified" setting.  I presume that the number in the path is a LUID for my external SD card on the old phone, and isn't valid on the new one.

Using a file explorer witth elevated permissions, if I browse into the Android/data/net.osmand.plus/fiiles directory, nothing is there.  (A net.osmand.plus folder exists under Android/obb, but it's completely empty.  There is no OsmAnd folder under Android/media.)

But my favorites are present in OsmAnd.  So wherre are they being stored?  SmartSync was able to copy data for other apps (not sure if all or not, I hate changing phones because it  takes a month to be sure everything is working correctly again), so why not for OsmAnd?

**Comment by vshcherb** [comment]
I think the message appears every time if OsmAnd doesn't have write permission to the folder, so it could read but it can't change anything. In that case it malfunction and probably gets stuck at some point. I think it's common Google policy that apps couldn't share same folders any more

**Reply by jfburke3** [reply]
@vshcherb OK, that kinda makes sense. So I just have to fix it manually.

**Reply by useperl** [reply]
What @jfburke3 wrote in https://github.com/osmandapp/OsmAnd/discussions/15348#discussion-4407665 (his comment of Sep 20, 2022) is a **perfect description of my** problem. I also migrated from an old Samsung device (tablet) to a new one (Tab S2 -> Tab S7FE). And this "evergreen" (always re-appearing) "Migrate shared storage" dialog had been my nightmare for a few months. Actually, I came here, searching for a remedy/solution. But as I found no solution in here (or anywhere), I improvised and seem to have found a solution myself.
I opened the `≡` (Menu under the triple horizontal bars in the bottom-left corner) -> ⚙️ `Settings` -> `OsmAnd settings` -> `Data storage folder`
Under `Change storage folder` section, `Shared memory` seems to have been selected, with a warning "_Shared memory is no longer available due to changed Android storage access_"
I selected `Internal app memory` (pushed the radio button on the right).
And **voilà** -- my nightmare seems to be over.

**Reply by endolith** [reply]
@useperl  That fixed it for me, too

**Reply by SebastianFloKa** [reply]
well, my nightmare seems to be over as well. Thank you sooo much for this fix @useperl

**Reply by netless-ww** [reply]
Well done useperl - fixed for me too 😊

**Reply by tbnorth** [reply]
Worked for be too, many thanks. Want to add that the feedback from the app. makes it sound like this is a big deal, I didn't want to lose my points and tracks, or leave wasted space with map downloads, but none of that happened and the problem is solved.

**Comment by ralhin** [comment]
I have almost exactly your problem, jfburke, only from a LG G5 (all data on a "external" memory card) to a google pixel 4a (not having any "external" memory) using google transfer. no more "nightmare" after selecting internal app memory. but as you I do see my favorites but miss very much my specific profiles I constructed over time and have been very happy with! the strangest thing is that it worked very well right from the start INCLUDING my old profiles, but (probably after setting to internal) these disappeared. but where? as you I cannot find (using google files app) ANY osmand data, in fact no internal data at all, these seem to be hidden. also not via osmand when trying to change data adress. but when connected to a pc I do see this: folder android>obb>net.osmand.plus (empty), data>net.osmand.plus>cache (empty), data>net.osmand.plus>files (including 4 folders: rendering including one zero byte file, routing = empty, tiles with folder osmand online tiles with one metainfo-file, tracks = empty and one 0-byte nomedia-file). where are my downloaded maps and my favorites? how can I get my old profiles back, which I already have used on the new phone so they should be somewhere?

btw: only by thorougly searching my new phone I discovered that its camera does produce also RAW-files, but in a different location (pictures) than the jpegs (DCIM>camera). can I make pixel store them all in the same place?

thank you!

**Comment by zzlazz** [comment]
One year after above messages ... Osmand paying version, exactly the same problem, migrating trough Google transfer from (lost) S9 with osmand  data on SD card and android 10 to fairphone 4 (android 12) without SD card. 
This is reaaaally annoying. Osmand  still usable but not nice. You can download maps, but every time you get to that *&%$#^^ screen about shared migration.
BTW have been unable to locate where Osmand stores its data on the new phone 
@sonora

**Comment by sonora** [comment]
What causes the issue seems quite clear: When some mechanism like Samsung SmartSwitch transfers an app's storage folder to a new device with a higher Android version.

The complexity of Android's newer storage policy makes a difference for what is "allowed" between
- existing app installs when upgrading Android on a device,
- updating an existing app install depending on the Android version, 
- new app installs (depending on the Android version).

Device transfer processes create yet another scenario, likely forcing the immediate storage migration requests described above.

Some things aggravate the situation and its investigation (transparency):
- not all folders are accessible any more in Android by file manager apps. As a consequence some may appear empty but are not.
- OsmAnd creates a number of bogus (ultimately unused) data folders (#18117)
- For some functionality like the favorites OsmAnd uses the corresponding files as a backup but for operations uses a database in memory (likely transferred ok by device switcher apps)..

Users who experience the issue described above should check in the Android permissions that OsmAnd has been granted the Storage permission. Other than that (and as long as we don't give OsmAnd 'All files access', #12046) I guess the solution can only be using OsmAnd's Data storage setting to pick an 'allowed' storage.

**Reply by zzlazz** [reply]
As far as I see neither osmand nor osmand+ can NOT even be allowed the storage permission on my Fairphone 4, recently updated to Android 13. I think it was the same on Android 12.
So I am still tapping "skip" twice when using osmand :-(

**Comment by sonora** [comment]
Perhaps users who continue to have an issue with this can try this procedure I have written up here https://osmand.net/docs/user/troubleshooting/maps-data/#picking-a-generally-accessible-osmand-data-storage-folder-using-the-media-storage and report if they have success with this.

**Comment by gb-dursley** [comment]
The same problem popped up on my Samsung Galaxy Tab S3 (Android 9) some weeks ago.  The nonsense instructions in the error defeated me - whatever is supposed to happen by default doesn't work!  OSMAnd had been running without this problem for 18 months and the only change (I think) is there had been a recent update to OSMAnd from F-Droid.
However, having separately pointed OSMAnd at the external folder already containing local maps, favourites etc. the problem went away.  Last night it returned.  The only thing that was different from the recent, fault-free, use was that I had the external storage (different folder) open in my file manager. Managed to do the simple 'how far, walking, is this from home?' task before shutting the tablet down. This morning problem still exists. OSM flags a crash, but the errorr reporting process doesn't work!!  However, after a second 'close all Apps and Restart' OSMAnd is working properly (at full speed) - using local resources and with all maps, favourites and my tracks available.  A quick check reveals that all my downloaded, but mostly disabled for speed, European maps are present.  Definitely a BUG. Whether that's in Android's file management or in OSM I'm in no position to guess, but I'd suggest that the problem is that OSM is seeing that all root folders including /Android are folder is open, rather than it's own sub-folders in /Android/data.
 
As a work-around, I suggest not having a File Manager running when launching OSMAnd.

**Reply by sonora** [reply]
Is it reproducible? Because it may also be a hardware issue the SD card...

**Reply by gb-dursley** [reply]
Since writing the above nearly 3 months ago, the problem has vanished; the following might help someone.

(OSM) Settings > Data Storage Folder shows:-

**Internal App Memory**

**External Storage 1**
   /storage/emulated/0/Android/data/net.osm.plus/files

**External Storage 2**
   /storage/F411-2B75/Android/data/net.osm.plus/files       [SELECTED]

**Multiuser  storage 1**
    /storage/emulated/0/Android/obb/net.osm.plus/files

**Multiuser  storage 2**
   /storage/F411-2B75/Android/obb/net.osm.plus/files

**Manually specified**
   /storage/F411-2B75/Android/data/net.osm.plus/files

Strangely, although both **External storage 2** and **Manually specified** point to the same physical location on the SD card, after changing to **External storage 2**, the problem has not reappeared.  On the sound engineering principle of 'If it ain't broke, don't fix it!' I haven't tried changing back.

I doubt the particular SD card is the problem; I store all my files on it and no other app has any problem accessing the card - other than those which simply cannot access external storage - surprisingly common, and not just confined to Android.

**Reply by sonora** [reply]
Probably you mean net.osm**and**.plus in all paths above.

Why 'manually specified' would not work as long as it points to one of the paths google "officially allows" (i.e. all the ones in the preconfigured options) beats me.

**Comment by prmartinuk** [comment]
Having experienced the same issue after moving to a new phone using the Android phone switch system I realised that the issue for me was that OSMAnd still had the path for my old phone's external SD card in the settings.
Going to Settings -> Data Storage and selecting any of the other options given for storage paths fixed the issue for me.
I didn't have many customisations so I'm not sure if any settings were lost or not migrated.

**Reply by sonora** [reply]
Yes, the message seems to generally be invoked when the already selected storage path is found no longer writeable for the app.

While a few years back the Android-mandated 'Shared storage migration' was the prime reason for that happening (use cases were an Android update on the device or an OsmAnd app update to a version with higher target SDK), there are nowadays mostly other special use cases like the a phone migration using e.g. Android switch.

Hemce probably nowadays the corresponding 'Shared storage migration'  message the app produces is no longer adequate and more often misleading than helpful.

@vshcherb  Perhaps a simple string change prompting to 'Select a 'writeable storage folder'  instead of still confusing people about a 'Shared storage migration' could make things clearer.
